pwndbg/gdblib/dynamic.py

```diff
@@ -0,0 +1,801 @@
+ld.so, and exposes an event that other parts of pwndbg can tap into, but one
+that may have a somewhat obtuse beahvior, due to limitations in GDB. See
+`r_debug_install_link_map_changed_hook` for more information.
+"""
+
+from __future__ import annotations
+
+import gdb
+
+import pwndbg.color.message as message
+import pwndbg.gdblib.bpoint
+import pwndbg.gdblib.memory
+import pwndbg.gdblib.typeinfo
+import pwndbg.lib.cache
+from pwndbg.lib.elftypes import constants as elf
+
+
+def _r_debug():
+    """
+    The easiest entry point into the link map is through the debug structure
+    provided by ld.so. It provides a convenient pointer into the head of the
+    link map list[1], and can be found at the address that corresponds to the
+    `_r_debug` symbol[2].
+
+    [1]: https://elixir.bootlin.com/glibc/latest/source/elf/link.h#L45
+    [2]: https://elixir.bootlin.com/glibc/glibc-2.38/source/elf/dl-debug-symbols.S#L30
+    """
+
+    try:
+        address = gdb.execute("output/x &_r_debug", to_string=True)
+        address = int(address, 0)
+
+        return address
+    except gdb.error:
+        # Symbol is most likely unavailable.
+        return None
+
+
+def is_dynamic():
+    """
+    Returns whether the current inferior is dynamic.
+
+    Not all programs are dynamically linked, or even need the dynamic loader at
+    all. Since this module is entirely reliant on at least the presence of the
+    dynamic loader, and really only makes sense for dynamic programs, it should
+    not be used at all with programs that don't participate in dynamic linkage,
+    or when there is a dynamic linker, but we have no way to talk to it.
+    """
+    return _r_debug() is not None
+
+
+# Reference to our hook in the link map update breakpoint, if it is installed.
+R_DEBUG_LINK_MAP_CHANGED_HOOK = None
+R_DEBUG_LINK_MAP_CHANGED_LISTENERS = set()
+
+
+class RDebugLinkMapChangedHook(pwndbg.gdblib.bpoint.BreakpointEvent):
+    """
+    Hook that gets activated whenever the link map changes.
+
+    The r_debug structure, in addition to having a refence to the head of the
+    link map, also has, in its ABI-stable part, a reference to an address that
+    can have a breakpoint attached to it, such that whenever the contents of the
+    link map change, that breakpoint will be triggered[1].
+
+    We take advantage of that here, by installing our own breakpoint in that
+    location, and watching for trigger events, so that we can notify other bits
+    of pwndbg that the contents of the `link_map()` function will be different.
+
+    [1]: https://elixir.bootlin.com/glibc/glibc-2.37/source/elf/link.h#L52
+    """
+
+    skip_this = True
+
+    def on_breakpoint_hit(self):
+        # Skip every other trigger, we only care about the completed link map
+        # that is available after the library is loaded.
+        if self.skip_this:
+            self.skip_this = False
+            return
+        else:
+            self.skip_ths = True
```

Comment:
Typo?

Also maybe just `self.skip_this = not self.skip_this`

Reply:
Yes, and I've fixed it. Thanks